chore: add internal stats endpoint

janniks opened the pull request:
- adds `/ordinals/stats/inscriptions` endpoint to show `count` of inscriptions genesis per block
- [ ] how do we typically add tests here?
- [ ] any types i need to be aware of?
- [ ] is it possible to hide the api from the public docs (these might be considered internal and not conforming to semver)
- [x] can these be cached and invalidated on each chainhook ingest?

sample response for `http://localhost:4000/ordinals/stats/inscriptions`

```
{"results":[[767430,"1","1"],[767753,"1","2"],[768094,"1","3"],[770110,"0","3"],[770530,"1","4"],[771290,"1","5"],[771573,"1","6"],[771694,"1","7"],[771717,"1","8"],[771971,"0","8"],[771993,"0","8"],[771994,"0","8"],[772088,"0","8"],[772119,"1","9"],[772167,"1","10"],[772642,"1","11"],[772657,"1","12"],[772676,"1","13"],[772679,"1","14"],[772731,"1","15"],[ ...
```

rafaelcr commented on src/pg/pg-store.ts:
```diff
@@ -317,4 +319,20 @@ export class PgStore extends BasePgStore {
+  async getInscriptionCountPerBlock(): Promise<ValuesRowList<DbInscriptionCountPerBlock[]>> {
+    return await this.sql<DbInscriptionCountPerBlock[]>`
+      SELECT t1.block_height,
+        COUNT(*) FILTER (WHERE t2.genesis = true) AS count,
+        SUM(COUNT(*) FILTER (WHERE t2.genesis = true)) OVER (ORDER BY t1.block_height) AS scan_count
```
I think it would be better to keep a table that accumulates these counts so we can perform these queries much faster (counts will get very heavy once the tables start growing larger).

The ingestion process needs to be adjusted for this to happen. We could do like a materialized view that we refresh with every new bitcoin block or something like that.

rafaelcr reviewed (commented):
I think this is looking good, but it would need tests before we can merge it. I don't think we'll be able to keep this endpoint private to us because the explorer is client-side, so we'll have to make sure these endpoints are as performant as possible

janniks commented:
Added a materialized view alternative. Unfortunately, postgres doesn't have partial refreshes or IVM yet. Which means, if we would maintain a table ourselves (and partially refresh/populate it), that would be much more efficient. But comes at the burden of maintenance/dev-cost.

I also added a debounced refresh (with JavaScripted timeouts). The debounce is the traditional trailing-edge debounce. It will wait for 10 seconds after the last event, if no further even is processed (i.e. idle for 10 seconds) then the refresh is called once.

rafaelcr commented on migrations/1681244251000_stats.ts:
```diff
@@ -0,0 +1,27 @@
+    },
+    `SELECT block_height,
+      COUNT(*) AS count,
+      SUM(COUNT(*)) OVER (ORDER BY block_height) AS scan_count
```
I'm not clear on what `scan_count` means in this context, could you explain a bit?

janniks replied:
I chose `scan` as the cumulative `reduce` algorithm. But it definitely sounds confusing! Going with `inscriptions` and `inscriptions_total` (for cumulative) in the new PR

rafaelcr commented on src/chainhook/server.ts:
```diff
@@ -125,6 +130,7 @@ const Chainhook: FastifyPluginCallback<Record<never, never>, Server, TypeBoxType
   fastify.post('/chainhook/inscription_revealed', async (request, reply) => {
     try {
       await processInscriptionRevealed(request.body, fastify.db);
+      refreshMaterializedViewsDebounced(fastify.db);
```
Chainhook messages come once per block, so it's not necessary to create this debounced logic. You could simply issue a refresh once per message.

rafaelcr reviewed (changes requested):
A couple observations. Also, could you rebase this branch to the latest `develop`? I've just added the latest changes into it. Thanks!

rafaelcr commented:
Thanks for the PR @janniks !

I think we might be able to get this result without a materialized view, though. Based on the last chainhook updates from Ludo, we get 1 message per block with all isncriptions inside of it. Which means, you could simply create a table that adds to the total inscriptions for that block height accordingly. You would subtract from this total if there is a rollback.

Thoughts?

janniks commented on src/api/routes/stats.ts:
```diff
@@ -0,0 +1,25 @@
+  fastify,
+  options,
+  done
+) => {
```
This isn't relevant any more, right?

janniks commented:
Closed in favor of #70
